src/dune_rules/expander.ml

```diff
@@ -156,45 +156,40 @@ let expand_version { scope; _ } ~(source : Dune_lang.Template.Pform.t) s =
 ;;
 
 let expand_artifact ~source t a s =
-  match t.lookup_artifacts with
-  | None -> isn't_allowed_in_this_position ~source
-  | Some lookup ->
-    let path = Path.Build.relative t.dir s in
-    let name = Path.Build.basename path in
-    let dir = Path.Build.parent_exn path in
-    let does_not_exist ~loc ~what name =
-      User_error.raise ~loc [ Pp.textf "%s %s does not exist." what name ]
+  let path = Path.Build.relative t.dir s in
+  let loc = Dune_lang.Template.Pform.loc source in
+  let name = Path.Build.basename path in
+  let dir = Path.Build.parent_exn path in
+  if Path.Build.is_root dir
+  then User_error.raise ~loc [ Pp.text "cannot escape the workspace root directory" ];
```

Comment:
Is this check as general as it could be? What happens if we write `%{cma:../../../foo}` in the test (so that we go way outside the root directory)?

Reply:
I added a test - It doesn't go through.

By the way, there's still a way to do something illegal here. One can write:

```
%{cma:../another-context/foo}
```

And use the relative path to escape to another context. We should definitely forbid that as well. It's a much larger change though.